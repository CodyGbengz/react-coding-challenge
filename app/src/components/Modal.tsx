import React from "react";
import styled from "styled-components";

import useTheme from "../hooks/useTheme";

const StyledModal = styled.section`
  border: 1px solid black;
  background: ${(props) => props.theme.primary};
  color: ${(props) => props.theme.text};
  display: flex;
  flex-flow: row wrap;
  height: 100vh;
  justify-content: space-between;
  padding: 4%;
  position: absolute;
  top: 0px;
  width: 85%;

  .imageWrapper {
    width: 45%;
  }

  .textWrapper {
    width: 45%;
  }
`;

const StyledOption = styled.option`
  text-align: center;
  width: 100px;
`;

export interface ModalProps {
    item: IItem;
    onClose: () => void;
    handleSelectStatus: (data: IItem) => void;
}

export const Modal: React.FC<ModalProps> = ({ item, handleSelectStatus }) => {
  const { theme } = useTheme();
  const statuses = ["error", "new", "processing", "done"];

  return item ? (
    <StyledModal onClick={(event) => event.stopPropagation()} theme={theme}>
      <div className="imageWrapper">
        <img
          src={item.image || "http://via.placeholder.com/300"}
          alt="image description"
        />
      </div>
      <div className="textWrapper">
        <h4>{item.name}</h4>
        <span>
          <select
            onChange={(event) =>
              handleSelectStatus({
                ...item,
                status: event.target.value as TItemStatus,
              })
            }
            name="status"
            id="status"
            defaultValue={item.status}
          >
            {statuses.map((status) => (
              <StyledOption
                key={status}
                value={status}
              >
                {status}
              </StyledOption>
            ))}
          </select>
        </span>
      </div>
    </StyledModal>
  ) : null;
};
