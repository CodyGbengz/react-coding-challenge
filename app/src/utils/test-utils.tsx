import React, {FC, ReactElement} from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import  theme  from '../configs/theme';
import { ThemeContext } from '../Root';

const AllTheProviders: FC = ({children}) => {
  return (
    <ThemeProvider theme={theme}>
        <ThemeContext.Provider value={{ theme }}>
            {children}
        </ThemeContext.Provider>
    </ThemeProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, {wrapper: AllTheProviders, ...options})

export * from '@testing-library/react'
export {customRender as render}