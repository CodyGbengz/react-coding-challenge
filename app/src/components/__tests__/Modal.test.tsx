import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom'

import { render } from '../../utils/test-utils'
import { Modal } from '../Modal';

describe('Modal', () => {
    const item:IItem = {
        id: 2,
        name: "Dragon",
        image: "https://picsum.photos/200/300",
        status: "new"
    };
    const onClose = jest.fn();
    const handleSelectStatus = jest.fn()
    it('should render correctly', () => {
        const { container } = render(
            <Modal 
                onClose={onClose}
                item={item}
                handleSelectStatus={handleSelectStatus}
            />
        )
        expect(screen.getByText('Dragon')).toBeInTheDocument();
        expect(container.parentElement).toMatchSnapshot();
    });
})
