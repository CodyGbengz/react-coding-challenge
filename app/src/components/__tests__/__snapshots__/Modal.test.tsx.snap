// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render correctly 1`] = `
<body>
  <div>
    <section
      class="sc-dkPtRN fijTDQ"
    >
      <div
        class="imageWrapper"
      >
        <img
          alt="image description"
          src="https://picsum.photos/200/300"
        />
      </div>
      <div
        class="textWrapper"
      >
        <h4>
          Dragon
        </h4>
        <span>
          <select
            id="status"
            name="status"
          >
            <option
              class="sc-hKwDye dyBJLE"
              value="error"
            >
              error
            </option>
            <option
              class="sc-hKwDye dyBJLE"
              selected=""
              value="new"
            >
              new
            </option>
            <option
              class="sc-hKwDye dyBJLE"
              value="processing"
            >
              processing
            </option>
            <option
              class="sc-hKwDye dyBJLE"
              value="done"
            >
              done
            </option>
          </select>
        </span>
      </div>
    </section>
  </div>
</body>
`;
